### ISSUE TITLE
Upon export and subsequent import of a zpool, the spare drives are being renamed on CentOS 7.6

### ISSUE THREAD

**Issue by 0n0w1c** [zfsonlinux/zfs#8439]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	|  Cent OS
Distribution Version	|  7.6.1810
Linux Kernel	|  3.10.0-957.5.1
Architecture	|  x86_64
ZFS Version	|  zfs-0.7.12-1
SPL Version	|  spl-0.7.12-1

### Describe the problem you're observing
Upon export and subsequent import of a zpool, the spare drives are being renamed on CentOS 7.6.

### Describe how to reproduce the problem
```
$ zpool create raid2z -f -o ashift=12 test /dev/disk/by-id/ata-CT250MX500SSD1_1905E1E7482F /dev/disk/by-id/ata-CT250MX500SSD1_1905E1E73DFD /dev/disk/by-id/ata-CT250MX500SSD1_1905E1E74833
$ zpool add test spare /dev/disk/by-id/ata-Samsung_SSD_850_EVO_250GB_S21NNXAGA84225B
$ zpool status
  pool: test
 state: ONLINE
  scan: none requested
config:

	NAME                                             STATE     READ WRITE CKSUM
	test                                             ONLINE       0     0     0
	  raidz2-0                                       ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E7482F              ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E73DFD              ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E74833              ONLINE       0     0     0
	spares
	  ata-Samsung_SSD_850_EVO_250GB_S21NNXAGA84225B  AVAIL   

$ zpool export test
$ zpool import test
[root@arraytest ~]# zpool status
  pool: test
 state: ONLINE
  scan: none requested
config:

	NAME                                 STATE     READ WRITE CKSUM
	test                                 ONLINE       0     0     0
	  raidz2-0                           ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E7482F  ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E73DFD  ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E74833  ONLINE       0     0     0
	spares
	  sdf                                AVAIL   

$ zpool export test
$ zpool import -d /dev/disk/by-id test
  pool: test
 state: ONLINE
  scan: none requested
config:

	NAME                                 STATE     READ WRITE CKSUM
	test                                 ONLINE       0     0     0
	  raidz2-0                           ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E7482F  ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E73DFD  ONLINE       0     0     0
	    ata-CT250MX500SSD1_1905E1E74833  ONLINE       0     0     0
	spares
	  wwn-0x5002538d406e1314             AVAIL   

$ ls -l /dev/disk/by-id
total 0
lrwxrwxrwx. 1 root root  9 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E73DFD -> ../../sdc
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E73DFD-part1 -> ../../sdc1
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E73DFD-part9 -> ../../sdc9
lrwxrwxrwx. 1 root root  9 Feb 20 10:34 ata-CT250MX500SSD1_1905E1E73E02 -> ../../sde
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 ata-CT250MX500SSD1_1905E1E73E02-part1 -> ../../sde1
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 ata-CT250MX500SSD1_1905E1E73E02-part9 -> ../../sde9
lrwxrwxrwx. 1 root root  9 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E7482F -> ../../sdb
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E7482F-part1 -> ../../sdb1
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E7482F-part9 -> ../../sdb9
lrwxrwxrwx. 1 root root  9 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E74833 -> ../../sdd
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E74833-part1 -> ../../sdd1
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 ata-CT250MX500SSD1_1905E1E74833-part9 -> ../../sdd9
lrwxrwxrwx. 1 root root  9 Feb 20 13:54 ata-Samsung_SSD_850_EVO_250GB_S21NNXAGA84225B -> ../../sdf
lrwxrwxrwx. 1 root root 10 Feb 20 13:54 ata-Samsung_SSD_850_EVO_250GB_S21NNXAGA84225B-part1 -> ../../sdf1
lrwxrwxrwx. 1 root root 10 Feb 20 13:54 ata-Samsung_SSD_850_EVO_250GB_S21NNXAGA84225B-part9 -> ../../sdf9
lrwxrwxrwx. 1 root root  9 Feb 19 12:51 ata-TEAC_DVD-ROM_DV28EV -> ../../sr0
lrwxrwxrwx. 1 root root  9 Feb 20 10:34 scsi-36001c230b78bf00023d38ad4b485a1b0 -> ../../sda
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 scsi-36001c230b78bf00023d38ad4b485a1b0-part1 -> ../../sda1
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 scsi-36001c230b78bf00023d38ad4b485a1b0-part2 -> ../../sda2
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 scsi-36001c230b78bf00023d38ad4b485a1b0-part3 -> ../../sda3
lrwxrwxrwx. 1 root root  9 Feb 19 12:51 usb-Dell_Virtual_CDROM_1028_123456-0:0 -> ../../sr1
lrwxrwxrwx. 1 root root  9 Feb 19 12:51 usb-Dell_Virtual_Floppy_1028_123456-0:0 -> ../../sdg
lrwxrwxrwx. 1 root root  9 Feb 20 13:54 wwn-0x5002538d406e1314 -> ../../sdf
lrwxrwxrwx. 1 root root 10 Feb 20 13:54 wwn-0x5002538d406e1314-part1 -> ../../sdf1
lrwxrwxrwx. 1 root root 10 Feb 20 13:54 wwn-0x5002538d406e1314-part9 -> ../../sdf9
lrwxrwxrwx. 1 root root  9 Feb 20 13:53 wwn-0x500a0751e1e73dfd -> ../../sdc
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 wwn-0x500a0751e1e73dfd-part1 -> ../../sdc1
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 wwn-0x500a0751e1e73dfd-part9 -> ../../sdc9
lrwxrwxrwx. 1 root root  9 Feb 20 10:34 wwn-0x500a0751e1e73e02 -> ../../sde
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 wwn-0x500a0751e1e73e02-part1 -> ../../sde1
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 wwn-0x500a0751e1e73e02-part9 -> ../../sde9
lrwxrwxrwx. 1 root root  9 Feb 20 13:53 wwn-0x500a0751e1e7482f -> ../../sdb
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 wwn-0x500a0751e1e7482f-part1 -> ../../sdb1
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 wwn-0x500a0751e1e7482f-part9 -> ../../sdb9
lrwxrwxrwx. 1 root root  9 Feb 20 13:53 wwn-0x500a0751e1e74833 -> ../../sdd
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 wwn-0x500a0751e1e74833-part1 -> ../../sdd1
lrwxrwxrwx. 1 root root 10 Feb 20 13:53 wwn-0x500a0751e1e74833-part9 -> ../../sdd9
lrwxrwxrwx. 1 root root  9 Feb 20 10:34 wwn-0x6001c230b78bf00023d38ad4b485a1b0 -> ../../sda
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 wwn-0x6001c230b78bf00023d38ad4b485a1b0-part1 -> ../../sda1
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 wwn-0x6001c230b78bf00023d38ad4b485a1b0-part2 -> ../../sda2
lrwxrwxrwx. 1 root root 10 Feb 20 10:34 wwn-0x6001c230b78bf00023d38ad4b485a1b0-part3 -> ../../sda3
```

**Comment by 0n0w1c** [comment]
It seems very similar to this closed ticket.
zfs changes back to /dev/sdx device names upon reboot #2944

**Comment by kpande** [comment]
this is a support request that's better served on [IRC or zfs-discuss](http://list.zfsonlinux.org). it looks like you're not using a `cachefile` or it's not in your initrd. additionally, you can set `ZPOOL_IMPORT_PATH=/dev/disk/by-id/` in some script or environment profile to ensure zpool import _always_ uses these names.

**Comment by kpande** [comment]
and, exporting a pool removes it from the cache file (as I understand) because you're telling the system "we're done with this pool". don't export it if you want to keep it around.

**Comment by kpande** [comment]
potentially #8108 for ata/wwn device names mixing up

**Comment by 0n0w1c** [comment]
I should add that a normal reboot without exporting via command line first, works as expected without issue. The proper device name is retained. However, a reboot does not "fix" the device name.

There is no mention of the spare drives in the cache file. Which to me, seems odd considering what I understand the purpose of the cache file to be.

What does a reboot do differently to export the array than the export from the command line?

**Comment by kpande** [comment]
rebooting *should not* export the pool

**Comment by 0n0w1c** [comment]
Confirmed, my cache file does persist through reboots... too bad it does not keep track of the spare drives for the array but I will assume there is a reason it does not. I will try using the vdev_id.conf file, I assume that will solve my issue.

**Comment by kpande** [comment]
you can modify the dracut-zfs module to use a defined search path including `/dev/disk/by-id` to see if this solves the issue along with PR #8440 - if so, we can modify dracut to import disks using their IDs.

**Comment by 0n0w1c** [comment]
Well, I tried the /etc/zfs/vdev_id.conf file... same results.

$ cat vdev_id.conf
#     by-id
#     name     fully qualified or base name of device link
alias a1s1       /dev/disk/by-id/ata-CT250MX500SSD1_1905E1E73DFD
alias a1s2       /dev/disk/by-id/ata-CT250MX500SSD1_1905E1E7482F
alias a1s3       /dev/disk/by-id/ata-CT250MX500SSD1_1905E1E74833
alias a1s4       /dev/disk/by-id/ata-CT250MX500SSD1_1905E1E73E02
alias a1s5       /dev/disk/by-id/ata-Samsung_SSD_850_EVO_250GB_S21NNXAGA84225B

$ zpool create -o ashift=12 test raidz2 /dev/disk/by-vdev/a1s1 /dev/disk/by-vdev/a1s2 /dev/disk/by-vdev/a1s3
$ zpool add -o ashift=12 test spare /dev/disk/by-vdev/a1s5

$ zpool status
  pool: test
 state: ONLINE
  scan: none requested
config:

	NAME        STATE     READ WRITE CKSUM
	test        ONLINE       0     0     0
	  raidz2-0  ONLINE       0     0     0
	    a1s1    ONLINE       0     0     0
	    a1s2    ONLINE       0     0     0
	    a1s3    ONLINE       0     0     0
	spares
	  a1s5      AVAIL   

errors: No known data errors
$ zpool export test
$ zpool status
no pools available

$ zpool import test
$ zpool status
  pool: test
 state: ONLINE
  scan: none requested
config:

	NAME        STATE     READ WRITE CKSUM
	test        ONLINE       0     0     0
	  raidz2-0  ONLINE       0     0     0
	    a1s1    ONLINE       0     0     0
	    a1s2    ONLINE       0     0     0
	    a1s3    ONLINE       0     0     0
	spares
	  sdf       AVAIL   

errors: No known data errors

**Comment by 0n0w1c** [comment]
> you can modify the dracut-zfs module to use a defined search path including `/dev/disk/by-id` to see if this solves the issue along with PR #8440 - if so, we can modify dracut to import disks using their IDs.

I have not ventured into dracut as this a data array comprised of SSDs which have locking enabled. I use hdparm to unlock each drive, perform a partprobe and then finally import the zpool.

**Comment by kpande** [comment]
when you say "unlock each drive" do you mean SCSI PR-3?

**Comment by 0n0w1c** [comment]
SCSI PR-3? That is something I am not familiar with. I am referring to the Trusted Computing OPAL 2.0 Self-Encrypting Drive locking.

$ hdparm --user-master u --security-unlock adfitech /dev/sdb security_password="xxxxxxxx"

/dev/sdb:
 Issuing SECURITY_UNLOCK command, password="xxxxxxxx", user=user

**Comment by kpande** [comment]
SCSI persistent reservations, I thought you were sharing the SSDs between two hosts. SCSI PR-3 allows one host to mark disk as 'in use' so that the other can not access and split brain. I was going to suggest that we add support for this into `dracut-zfs` but what you describe is very different and would deserve its own dracut module.

**Comment by richardelling** [comment]
Spares and cache devices are internally known as auxiliary devices and, in the case of spares, were originally designed to be shareable amongst multiple pools. This means that though they are top-level vdevs, it is ok to import the pool without them: as in the case where a spare is in-use by another pool.

The pool does not store its entire configuration in the on-disk labels. Each top-level vdev configuration stored in the label of its leaf vdevs. For aux devices, there is only one leaf vdev, itself.

The label on disk for the aux devices does not contain a pool guid. So the method to import them is to compute the known top-level vdev guids then compare to the pool guid. If there is a mismatch then other disks are scanned to see if they are aux devices and if so, does their guid contribute to solving the mismatch.

Bottom line: to make aux devices act like other top-level vdevs is a fairly significant amount of design work that might conflict with previous design requirements. Is it worth the effort? I'm not convinced because once a spare is activated in a pool, then it becomes a real leaf vdev and has a full label.

**Comment by kpande** [comment]
@richardelling my approach was to modify dracut or other import routines to do the right thing here instead of expecting the cache file to describe things correctly.

**Comment by richardelling** [comment]
Those are bandages, which is fine for now. Long term aux devices can go away.
